File: backend/app/routers/data_stores.py
# backend/app/routers/data_stores.py
from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import get_current_active_user
from app.models import User
from app.repositories.data_store_repository import DataStoreRepository
from app.schemas.data_store import (
    DataStoreCreate,
    DataStoreUpdate,
    DataStoreResponse,
    DataStoreRowResponse,
    DataStoreRowsResponse,
    CsvImportResponse,
    ExportPreviewRequest,
    ExportPreviewResponse,
    ExportExecuteResponse,
)
from app.services.data_store_service import DataStoreService
from app.services.exceptions import ConflictError, NotFoundError, ValidationError

router = APIRouter(
    prefix="/projects/{project_id}/data-stores",
    tags=["data-stores"],
)


def get_data_store_service(db: AsyncSession = Depends(get_db)) -> DataStoreService:
    repo = DataStoreRepository(db)
    return DataStoreService(repo)


# ── Store CRUD ─────────────────────────────────────────────────────

@router.post("", response_model=DataStoreResponse, status_code=status.HTTP_201_CREATED)
async def create_data_store(
    project_id: UUID,
    data: DataStoreCreate,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Create a new data store for the project."""
    try:
        return await service.create_store(project_id, data)
    except ConflictError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("", response_model=list[DataStoreResponse])
async def list_data_stores(
    project_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """List all data stores for the project."""
    return await service.list_stores(project_id)


@router.get("/{store_id}", response_model=DataStoreResponse)
async def get_data_store(
    project_id: UUID,
    store_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Get a data store by ID."""
    try:
        return await service.get_store(store_id, project_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.patch("/{store_id}", response_model=DataStoreResponse)
async def update_data_store(
    project_id: UUID,
    store_id: UUID,
    data: DataStoreUpdate,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Update a data store."""
    try:
        return await service.update_store(store_id, project_id, data)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ConflictError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.delete("/{store_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_data_store(
    project_id: UUID,
    store_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Delete a data store and its underlying table."""
    try:
        await service.delete_store(store_id, project_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


# ── Row CRUD ───────────────────────────────────────────────────────

@router.get("/{store_id}/rows", response_model=DataStoreRowsResponse)
async def list_rows(
    project_id: UUID,
    store_id: UUID,
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """List paginated rows from a data store."""
    try:
        return await service.get_rows(store_id, project_id, limit, offset)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.post("/{store_id}/rows", response_model=DataStoreRowResponse, status_code=status.HTTP_201_CREATED)
async def insert_row(
    project_id: UUID,
    store_id: UUID,
    data: dict,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Insert a single row into a data store."""
    try:
        return await service.insert_row(store_id, project_id, data)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/{store_id}/rows/{row_id}", response_model=DataStoreRowResponse)
async def get_row(
    project_id: UUID,
    store_id: UUID,
    row_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Get a single row by ID."""
    try:
        return await service.get_row(store_id, project_id, row_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.patch("/{store_id}/rows/{row_id}", response_model=DataStoreRowResponse)
async def update_row(
    project_id: UUID,
    store_id: UUID,
    row_id: UUID,
    data: dict,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Update a single row."""
    try:
        return await service.update_row(store_id, project_id, row_id, data)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.delete("/{store_id}/rows/{row_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_row(
    project_id: UUID,
    store_id: UUID,
    row_id: UUID,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Delete a single row."""
    try:
        await service.delete_row(store_id, project_id, row_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


# ── CSV Import ─────────────────────────────────────────────────────

@router.post("/{store_id}/import", response_model=CsvImportResponse)
async def import_csv(
    project_id: UUID,
    store_id: UUID,
    file: UploadFile = File(...),
    column_mapping: str = Form(...),
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Import rows from a CSV file.

    column_mapping is a JSON string: {"csv_header": "store_column", ...}
    """
    import json

    if not file.filename or not file.filename.endswith('.csv'):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="File must be a CSV file"
        )

    try:
        mapping = json.loads(column_mapping)
    except json.JSONDecodeError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="column_mapping must be valid JSON"
        )

    content = await file.read()
    csv_text = content.decode("utf-8")

    try:
        return await service.import_csv(store_id, project_id, csv_text, mapping, filename=file.filename)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# ── Export Preview/Execute ────────────────────────────────────────

@router.post("/{store_id}/preview-export", response_model=ExportPreviewResponse)
async def preview_export(
    project_id: UUID,
    store_id: UUID,
    data: ExportPreviewRequest,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Preview export: validate mapping and return flattened rows without inserting."""
    try:
        return await service.preview_export(store_id, project_id, data.source_data, data.field_mapping)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/{store_id}/execute-export", response_model=ExportExecuteResponse)
async def execute_export(
    project_id: UUID,
    store_id: UUID,
    data: ExportPreviewRequest,
    current_user: User = Depends(get_current_active_user),
    service: DataStoreService = Depends(get_data_store_service),
):
    """Execute export: validate, flatten, and insert rows into the data store."""
    try:
        count = await service.execute_export(store_id, project_id, data.source_data, data.field_mapping)
        return ExportExecuteResponse(rows_imported=count)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

File: frontend/src/api/dataStores.ts
// frontend/src/api/dataStores.ts
import apiClient from './client'
import type {
  DataStore,
  DataStoreCreate,
  DataStoreUpdate,
  DataStoreRow,
  DataStoreRowsResponse,
  CsvImportResponse,
  ExportPreviewRequest,
  ExportPreviewResponse,
  ExportExecuteResponse,
} from '@/types/dataStore'

// ── Store CRUD ────────────────────────────────────────────────────

export async function listDataStores(projectId: string): Promise<DataStore[]> {
  const response = await apiClient.get<DataStore[]>(
    `/projects/${projectId}/data-stores`
  )
  return response.data
}

export async function getDataStore(projectId: string, storeId: string): Promise<DataStore> {
  const response = await apiClient.get<DataStore>(
    `/projects/${projectId}/data-stores/${storeId}`
  )
  return response.data
}

export async function createDataStore(projectId: string, data: DataStoreCreate): Promise<DataStore> {
  const response = await apiClient.post<DataStore>(
    `/projects/${projectId}/data-stores`,
    data
  )
  return response.data
}

export async function updateDataStore(
  projectId: string,
  storeId: string,
  data: DataStoreUpdate
): Promise<DataStore> {
  const response = await apiClient.patch<DataStore>(
    `/projects/${projectId}/data-stores/${storeId}`,
    data
  )
  return response.data
}

export async function deleteDataStore(projectId: string, storeId: string): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/data-stores/${storeId}`)
}

// ── Row CRUD ──────────────────────────────────────────────────────

export async function listRows(
  projectId: string,
  storeId: string,
  limit = 50,
  offset = 0
): Promise<DataStoreRowsResponse> {
  const response = await apiClient.get<DataStoreRowsResponse>(
    `/projects/${projectId}/data-stores/${storeId}/rows`,
    { params: { limit, offset } }
  )
  return response.data
}

export async function getRow(
  projectId: string,
  storeId: string,
  rowId: string
): Promise<DataStoreRow> {
  const response = await apiClient.get<DataStoreRow>(
    `/projects/${projectId}/data-stores/${storeId}/rows/${rowId}`
  )
  return response.data
}

export async function insertRow(
  projectId: string,
  storeId: string,
  data: Record<string, unknown>
): Promise<DataStoreRow> {
  const response = await apiClient.post<DataStoreRow>(
    `/projects/${projectId}/data-stores/${storeId}/rows`,
    data
  )
  return response.data
}

export async function updateRow(
  projectId: string,
  storeId: string,
  rowId: string,
  data: Record<string, unknown>
): Promise<DataStoreRow> {
  const response = await apiClient.patch<DataStoreRow>(
    `/projects/${projectId}/data-stores/${storeId}/rows/${rowId}`,
    data
  )
  return response.data
}

export async function deleteRow(
  projectId: string,
  storeId: string,
  rowId: string
): Promise<void> {
  await apiClient.delete(
    `/projects/${projectId}/data-stores/${storeId}/rows/${rowId}`
  )
}

// ── CSV Import ────────────────────────────────────────────────────

export async function importCsv(
  projectId: string,
  storeId: string,
  file: File,
  columnMapping: Record<string, string>
): Promise<CsvImportResponse> {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('column_mapping', JSON.stringify(columnMapping))

  const response = await apiClient.post<CsvImportResponse>(
    `/projects/${projectId}/data-stores/${storeId}/import`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  )
  return response.data
}

// ── Export Preview/Execute ────────────────────────────────────────

export async function previewExport(
  projectId: string,
  storeId: string,
  data: ExportPreviewRequest
): Promise<ExportPreviewResponse> {
  const response = await apiClient.post<ExportPreviewResponse>(
    `/projects/${projectId}/data-stores/${storeId}/preview-export`,
    data
  )
  return response.data
}

export async function executeExport(
  projectId: string,
  storeId: string,
  data: ExportPreviewRequest
): Promise<ExportExecuteResponse> {
  const response = await apiClient.post<ExportExecuteResponse>(
    `/projects/${projectId}/data-stores/${storeId}/execute-export`,
    data
  )
  return response.data
}
